src/net/socket_utils.zig

```diff
@@ -18,74 +19,344 @@ pub const PACKETS_PER_BATCH: usize = 64;
+            std.debug.assert(ref_count.swap(1, .acquire) == 0);
+            defer std.debug.assert(ref_count.swap(0, .release) == 1);
+
+            notifyIoThread(); // wake up xev thread to see ref_count == 1 to stop/shutdown
+            io_thread.join();
+            io_event.deinit();
+        }
+
+        st.handle.wait();
     }
 
-    fn spawn(
-        allocator: Allocator,
-        logger: Logger,
-        socket: UdpSocket,
-        channel: *Channel(Packet),
-        exit: ExitCondition,
-        comptime runFn: anytype,
-    ) !*SocketThread {
-        // TODO(king): store event-loop data in SocketThread (hence, heap-alloc)..
-        const self = try allocator.create(SocketThread);
-        errdefer allocator.destroy(self);
+    fn notifyIoThread() void {
+        const already_notified = io_notified.swap(true, .release);
+        if (!already_notified) {
+            io_event.notify() catch std.debug.panic("failed to notify xev event loop", .{});
+        }
+    }
 
-        self.* = .{
-            .allocator = allocator,
-            .handle = try std.Thread.spawn(.{}, runFn, .{ logger, socket, channel, exit }),
-        };
+    fn runIoThread() !void {
+        var loop = try xev.Loop.init(.{});
+        defer loop.deinit();
 
-        return self;
+        // Trigger pollIo() when the xev thread is notified from outside (channel send, shutdown).
+        var io_event_completion: xev.Completion = undefined;
+        io_event.wait(&loop, &io_event_completion, void, null, onNotify);
+
+        // Every once in a while, trigger pollIo() to check SocketThread.exit flags.
+        var timer = try xev.Timer.init();
+        defer timer.deinit();
+        var timer_completion: xev.Completion = undefined;
+        timer.run(&loop, &timer_completion, SOCKET_TIMEOUT_US, xev.Timer, &timer, onTick);
+
+        try loop.run(.until_done);
     }
 
-    pub fn join(self: *SocketThread) void {
-        self.handle.join();
-        self.allocator.destroy(self);
+    fn onNotify(
+        _: ?*void,
+        loop: *xev.Loop,
+        completion: *xev.Completion,
+        result: xev.Async.WaitError!void,
+    ) xev.CallbackAction {
+        result catch |e| std.debug.panic("xev notify event failed: {}", .{e});
+        std.debug.assert(io_notified.swap(false, .acquire));
+
+        pollIo(loop);
+        io_event.wait(loop, completion, void, null, onNotify);
+        return .disarm;
     }
 
-    fn runReceiver(
-        logger_: Logger,
-        socket_: UdpSocket,
-        incoming_channel: *Channel(Packet),
-        exit: ExitCondition,
+    fn onTick(
+        maybe_timer: ?*xev.Timer,
+        loop: *xev.Loop,
+        completion: *xev.Completion,
+        result: xev.Timer.RunError!void,
+    ) xev.CallbackAction {
+        result catch |e| std.debug.panic("xev notify event failed: {}", .{e});
+
+        pollIo(loop);
+        maybe_timer.?.run(loop, completion, SOCKET_TIMEOUT_US, xev.Timer, maybe_timer, onTick);
```

Comment:
We take the timer as an optional, and then almost immediately assert it isn't

Reply:
Okay, seems the optional is required to satisfy libxev's types? In this case, maybe just not use the "maybe_" param names on functions where we actually require it.

Reply:
https://github.com/Syndica/sig/pull/465/commits/693219d352ad864800c5ad9da20211113ba6afab